### DISCUSSION TITLE
Is it parallel? Can it grep through .zips and .jars?

### DISCUSSION THREAD

**Discussion by vemv** [dainiusjocas/lucene-grep#215] [General]
Hey there again, a few questions:

* Is search parallel across files?
* Does it find files through .zip and .jar archives?
* Let's say it goes through .zip files. How is parallelism distributed?
  * The info I'm seeking is if an optimal amount of threads is pursued. e.g. both "one thread per file" and "one thread per .zip" can be suboptimal.

I'm willing to contribute ideas/PRs if there happened to be a need.

Cheers - V

**Comment by dainiusjocas** [comment]
Hi, thanks for the interest! 

> Is search parallel across files?

No, only across lines of text. In my tests the overall throughput was higher with this setup.

> Does it find files through .zip and .jar archives?

Well, you can throw binary files at lmgrep but the results are unpredictable. To support .zip and .jar files it would require to detect the format, and open streams to text files in the archive. Should be doable.

> Let's say it goes through .zip files. How is parallelism distributed?

Parallelism currently works like this: iterating through files sequentially, reading lines of text and pushing them onto the thread-pool, the results (if any) are pushed onto another single-threaded thread pool for printing to the stdout.

> The info I'm seeking is if an optimal amount of threads is pursued. e.g. both "one thread per file" and "one thread per .zip" can be suboptimal.

Currently, the optimal amount of threads for greping should be ` total-cores-available - 1` where one thread is responsible for the printing the results. In my previous attempts, the parallelism per lines instead of per files gave better throughput. Also, file by file approach makes it "easier" to do command line tool.

**Reply by vemv** [reply]
> No, only across lines of text. In my tests the overall throughput was higher with this setup.

SGTM!

> To support .zip and .jar files it would require to detect the format, and open streams to text files in the archive. Should be doable.

Would be much appreciated, it's basically my use case. I could try contributing it as a PR. Feel free to @ me in an issue.

**Comment by dainiusjocas** [comment]
I'm planing to extract a library for greping not so far in the future. 

What is missing for the library is two big things:
- support multiple fields per document
- decide on the index mapping format
